azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/error/ChatErrorHandler.kt

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.android.communication.ui.chat.error
+
+import com.azure.android.communication.ui.chat.configuration.ChatCompositeConfiguration
+import com.azure.android.communication.ui.chat.models.ChatCompositeErrorCode
+import com.azure.android.communication.ui.chat.models.ChatCompositeErrorEvent
+import com.azure.android.communication.ui.chat.redux.AppStore
+import com.azure.android.communication.ui.chat.redux.state.ReduxState
+import com.azure.android.communication.ui.chat.utilities.CoroutineContextProvider
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.launch
+internal class ChatErrorHandler(
+    coroutineContextProvider: CoroutineContextProvider,
+    private val store: AppStore<ReduxState>,
+    private val configuration: ChatCompositeConfiguration,
+) {
+    private val coroutineScope = CoroutineScope((coroutineContextProvider.Default))
+    private var lastChatErrorEvent: ChatCompositeErrorEvent? = null
+
+    fun start() {
+        coroutineScope.launch(Dispatchers.Default) {
+            store.getStateFlow().collect {
+                onStateChanged(it)
+            }
+        }
+    }
+    private fun onStateChanged(state: ReduxState) {
+        checkIfCallStateErrorIsNewAndNotify(
+            state.errorState.chatCompositeErrorEvent,
+            lastChatErrorEvent,
+        ) { lastChatErrorEvent = it }
+    }
+
+    private fun chatStateErrorCallback(chatStateError: ChatCompositeErrorEvent) {
+        try {
+            configuration.eventHandlerRepository.getOnErrorHandlers()
+                .forEach { it.handle(chatStateError) }
+        } catch (error: Throwable) {
+            // suppress any possible application errors
+        }
+    }
+
+    private fun checkIfCallStateErrorIsNewAndNotify(
+        newChatErrorEvent: ChatCompositeErrorEvent?,
+        lastChatErrorEvent: ChatCompositeErrorEvent?,
+        function: (ChatCompositeErrorEvent) -> Unit,
+    ) {
+        if (newChatErrorEvent != null && newChatErrorEvent != lastChatErrorEvent) {
+            if (shouldNotifyError(newChatErrorEvent)) {
+                function(newChatErrorEvent)
+                chatStateErrorCallback(newChatErrorEvent)
+            }
+        }
+    }
+
+    // TODO: Check the logic again when we need to expose more error
+    private fun shouldNotifyError(newCallStateError: ChatCompositeErrorEvent) =
```

Comment:
I do not think we should exclude those JOIN_FAILED and TOKEN_EXPIRED.
Is there a particular requirement? Unlike Call composite we do not have a setup screen

Reply:
We have the requirement to expose the Join failed and token expired for contoso for public review.

And for our current demo app, it would show the error alert after we press the launch button, before choose partial view/full screen.